src/array.jl

```diff
@@ -804,39 +804,57 @@ function levels!(A::CategoricalArray{T, N, R}, newlevels::Vector;
     A
 end
 
-function _unique(::Type{S},
-                 refs::AbstractArray{T},
-                 pool::CategoricalPool) where {S, T<:Integer}
-    nlevels = length(levels(pool)) + 1
-    order = fill(0, nlevels) # 0 indicates not seen
-    # If we don't track missings, short-circuit even if none has been seen
-    count = S >: Missing ? 0 : 1
-    @inbounds for i in refs
-        if order[i + 1] == 0
-            count += 1
-            order[i + 1] = count
-            count == nlevels && break
+# return unique refs (each value is unique) in the order of appearance in `refs`
+# similar to fallback Base.unique() implementation,
+# but short-circuits once references to all levels are encountered
+function _uniquerefs(A::AbstractCategoricalArray{T}) where T
+    arefs = refs(A)
+    res = similar(arefs, 0)
+    nlevels = length(levels(A))
+    seen = fill(false, nlevels+1) # +1 for 0 (missing ref)
+    maxunique = nlevels + (T >: Missing ? 1 : 0)
+    @inbounds for ref in arefs
+        if !seen[ref + 1]
+            push!(res, ref)
```

Comment:
Instead of pushing here, wouldn't it be more efficient to do `res = (1:maxunique)[seen]` at the end?

Reply:
That would not keep the order of the first appearance of references in `A`.